src/Lecture4.hs

```diff
@@ -257,7 +352,14 @@ the file doesn't have any products.
 -}
 
 calculateStats :: String -> String
-calculateStats = error "TODO"
+calculateStats = statsOrError . rowsToNonEmpty . mapMaybe parseRow . lines
+
+statsOrError :: Maybe (NonEmpty Row) -> String
+statsOrError = maybe "the file doesn't have any products" (displayStats . combineRows)
+
+rowsToNonEmpty :: [Row] -> Maybe (NonEmpty Row)
+rowsToNonEmpty []      = Nothing
+rowsToNonEmpty (x: xs) = Just $ x :| xs
```

Comment:
You don't need to write this function, you can use standard function [nonEmpty](https://hackage.haskell.org/package/base-4.16.0.0/docs/Data-List-NonEmpty.html#v:nonEmpty) that converts an ordinary list to optional `NonEmpty` list

Reply:
nice! thx